packages/browser-wallet/src/shared/utils/token-helpers.ts

```diff
@@ -101,28 +102,60 @@ export function getTokenUrl(
     });
 }
 
+function confirmMetadataUrl(field?: MetadataUrl) {
+    if (field && !field.url) {
+        throw new Error('Url field was present but did no contain an url');
+    }
+}
+
 /**
  * Fetches token metadata from the given url
- * TODO: add/improve validation of metadata structure.
  */
-export async function getTokenMetadata(tokenUrl: string): Promise<TokenMetadata> {
-    // TODO remove this hack, for production, or just when we have a proper collection for testing (with online metadata).
-    if (tokenUrl.includes('example')) {
-        return {
-            name: 'Wrapped CCD Token',
-            symbol: 'wCCD',
-            decimals: 6,
-            description: 'A CIS2 token wrapping the Concordium native token (CCD)',
-            thumbnail: { url: 'https://proposals.concordium.software/_static/concordium-logo-black.svg' },
-            display: { url: 'https://proposals.concordium.software/_static/concordium-logo-black.svg' },
-            artifact: { url: 'https://proposals.concordium.software/_static/concordium-logo-black.svg' },
-        };
+export async function getTokenMetadata(tokenUrl: string, network: NetworkConfiguration): Promise<TokenMetadata> {
+    // TODO remove this hack when we have a proper collection for testing (with online metadata).
+    if (!isMainnet(network) && tokenUrl.includes('https://some.example/token/wccd')) {
+        return WCCD_METADATA;
     }
+
     const resp = await fetch(tokenUrl, { headers: new Headers({ 'Access-Control-Allow-Origin': '*' }), mode: 'cors' });
     if (!resp.ok) {
         throw new Error(`Something went wrong, status: ${resp.status}`);
     }
-    return resp.json();
+
+    const {
+        name,
+        symbol,
+        decimals,
+        unique,
+        description,
+        thumbnail,
+        display,
+        artifact,
+        assets,
+        attributes,
+        localization,
+    } = await resp.json();
+
+    if (decimals && Number.isNaN(decimals)) {
+        throw new Error('Metadata contains incorrect decimals format');
+    }
+    confirmMetadataUrl(thumbnail);
+    confirmMetadataUrl(display);
+    confirmMetadataUrl(artifact);
+
+    return {
+        name,
+        symbol,
+        decimals,
+        unique,
+        description,
+        thumbnail,
+        display,
+        artifact,
+        assets,
+        attributes,
+        localization,
+    };
```

Comment:
Seems a bit much? Maybe this instead:

```suggestion
    const response = await resp.json();
    const { decimals, thumbnail, display, artifact } = response;

    if (decimals && Number.isNaN(decimals)) {
        throw new Error('Metadata contains incorrect decimals format');
    }
    confirmMetadataUrl(thumbnail);
    confirmMetadataUrl(display);
    confirmMetadataUrl(artifact);

    return response;
```

Reply:
Well I did it that way to filter other fields, in case someone added more :thinking:

Reply:
I guess I can do that more succinctly with a filter

Reply:
I ended reverting the filtering behaviour.